api/common/sharing/validators.go

```diff
@@ -0,0 +1,56 @@
+package sharing
+
+import (
+	"context"
+	"github.com/Peripli/service-manager/pkg/instance_sharing"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/tidwall/gjson"
+)
+
+func ValidateReferencedInstance(body []byte, tenantIdentifier string, repository storage.Repository, ctx context.Context, getTenantId func() string) (string, error) {
+	parameters := gjson.GetBytes(body, "parameters").Map()
+	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+
+	if !exists {
+		return "", util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+
+	referencedInstanceObj, err := storage.GetObjectByField(ctx, repository, types.ServiceInstanceType, "id", referencedInstanceID.String())
+	if err != nil {
+		log.C(ctx).Errorf("Failed retrieving the reference-instance by the ID: %s", referencedInstanceObj)
+		return "", util.HandleStorageError(util.ErrNotFoundInStorage, types.ServiceInstanceType.String())
+	}
+	referencedInstance := referencedInstanceObj.(*types.ServiceInstance)
+
+	//validate Ownership in case of a multi tenant flow
+	if tenantIdentifier != "" {
+		targetInstanceTenantID := referencedInstance.Labels[tenantIdentifier][0]
+		if targetInstanceTenantID != getTenantId() {
+			log.C(ctx).Errorf("Instance owner %s is not the same as the caller %s", targetInstanceTenantID, getTenantId())
```

Comment:
add target instance ID to log

Reply:
Hey Eyal,

I debugged the code and wrote some tests to cover it.
We've added to the `GetObjectByField` the `query.CriteriaForContext(ctx)` when we retrieve instances.

Doesn't it cover the ownership as well?

For example, when I provision a reference instance and retrieve the target instance (the shared instance we point to) via the `GetObjectByField` function, should I or shouldn't I be able to retrieve it?

Due to the `query.CriteriaForContext(ctx)`

If I should not be able to, due to the new criteria - we can consider removing the additional validation.

Thanks.

Reply:
Added an additional query `byLabel`
So we retrieve the shared instance by the `reference_instance_id` and the `callerTenant`
`byLabel := query.ByLabel(query.EqualsOperator, tenantIdentifier, getTenantId())`

Reply:
@eyalnathan - additional tests for ownership validation added:

1. `When("shared instance owned by different tenant")`
2. `When("tenant without ownership of instance is sharing instance")`